### PR TITLE
Assign categories to categorical columns

### LINKED ISSUE #800: Assign categories to categorical columns in TPP schema
The existing TPP schema contains columns which are logically categorical but we do not have a `CategoricalConstraint` applied to them. This means that Data Builder won't correctly encode them as categoricals when producing binary output, and will also struggle to generate dummy data for them.

Examples include (but are not limited to):
https://github.com/opensafely-core/databuilder/blob/39bc9fdd2895020b50b341b3a4957f43d55df2f7/databuilder/tables/beta/tpp.py#L35-L36
https://github.com/opensafely-core/databuilder/blob/39bc9fdd2895020b50b341b3a4957f43d55df2f7/databuilder/tables/beta/tpp.py#L81

An example of a column which _does_ have a correctly applied `CategoricalConstraint` is:
https://github.com/opensafely-core/databuilder/blob/39bc9fdd2895020b50b341b3a4957f43d55df2f7/databuilder/contracts/universal.py#L26-L36

### PR DESCRIPTION
**Pull request by iaindillingham**
I've reworked this PR. Originally, it added several categorical constraints that reflected the current state of the OpenSAFELY TPP database ❄️. Now, it adds one categorical constraint and two regex constraints that do not. Because I'm not confident that I can write a regex correctly, I also added three tests 🧞:

1. test columns with regex constraints against valid examples;
2. test columns with regex constraints against invalid examples;
3. test all columns with regex constraints are tested against valid and invalid examples.

For evidence that that constraints added in this PR hold with respect to the data in the OpenSAFELY TPP database, then please see the released outputs of the [tpp-database-categorical-columns][1] workspace.

[1]: https://jobs.opensafely.org/opensafely-internal/tpp-database-categorical-columns/

### CODE REVIEW

**Review by evansd — COMMENTED** [review]
Sorry Iain, this is  bit of downer review because I'm not actually sure _any_ of those columns should be categorical!

This is still a valuable piece of work though because we now know we've checked all the candidates for categorical columns. I think when the ticket was originally written we had very few categoricals but we must have covered the remaining ones in the intervening period as we went along.

And adding the regex constraint will immediately improve the dummy data here as well.

On the `NotNull` constraints, I agree with all the changes you've made here. But just to sound a note of caution that just because the database column prohibits NULLs in a literal sense, it doesn't mean TPP haven't encoded "semantic NULLs" in there somehow (`0`, `-1`, `9999-12-31` etc.). So I think we should only add that constraint if we're reasonably confident that the column will always be populated with a meaningful value.

**Comment by iaindillingham** [comment]
You asked, "Should any of these columns be categorical?" (where the columns are `addresses.address_type`, `ec.sus_hrg_code`, `opa.hrg_code`, `vaccinations.target_disease`) Each of these columns is an instance of the "non-exhaustive categorical constraints" type I mentioned above. We don't use `CASE` in the backend table, so we get what TPP give us. There are other instances, though. Specifically:

* `opa.attendance_status`
* `opa.consultation_medium_used`
* `opa.first_attendance`
* `practice_registrations.practice_nuts1_region_name`
* `sgss_covid_all_tests.variant_detection_method`
* `wl_clockstops.waiting_list_type`
* `wl_openpathways.waiting_list_type`
* `ethnicity_from_sus.code`

Should I change the nature of the constraints in these instances, too?

File: ehrql/tables/beta/tpp.py [hunk 1]
```diff
@@ -637,6 +637,21 @@ class opa(EventFrame):
             "The Healthcare Resource Group (HRG) code assigned to the activity, "
             "used to assign baseline tariff costs."
         ),
+        constraints=[
+            Constraint.Categorical(
```
[the remark this answers is not in the document]
Comment (iaindillingham) [reply ehrql/tables/beta/tpp.py:641]: I've replaced this with a `Constraint.Regex`.

File: ehrql/tables/beta/tpp.py [hunk 2]
```diff
@@ -1076,6 +1076,58 @@ class vaccinations(EventFrame):
     target_disease = Series(
         str,
         description="Vaccine's target disease.",
+        constraints=[
+            Constraint.NotNull(),
+            Constraint.Categorical(
```
[the remark this answers is not in the document]
Comment (iaindillingham) [reply ehrql/tables/beta/tpp.py:1081]: I've removed these constraints. I'd like to revisit adding a link to the output from [opensafely/tpp-database-categorical-columns](https://github.com/opensafely/tpp-database-categorical-columns), when I've confirmed that we can publish that outputs.

File: tests/unit/test_tables.py [hunk 3]
```diff
@@ -12,3 +17,69 @@ def test___all__(module):
...
+    "examples",
+    [valid_examples_for_regex_constraints, invalid_examples_for_regex_constraints],
+)
+def test_all_regex_constraints_are_tested(examples):
+    tables = {
+        table
+        for _, table in itertools.chain.from_iterable(
+            get_tables_from_namespace(module) for module in get_submodules(ehrql.tables)
+        )
+    }
```
[the remark this answers is not in the document]
Comment (iaindillingham) [reply tests/unit/test_tables.py:69]: Much nicer! I've avoided.

**Pull request merged**